### PR TITLE
Feature/identity middleware

### PR DESCRIPTION
**Pull request by CarlHembrough**
### What

Update identity middleware to call new identity endpoint and store response in request context.

* Added tests around existing functionality
* Read `Authorization` header from requests as well as existing Florence token
* If both headers are provided then the user header takes precedence
* Store `User-Identity` and `Caller-Identity` in request context to use in downstream calls.

### How to review

Review changes

### Who can review

Anyone

### CODE REVIEW

File: identity/identity.go [hunk 1]
```diff
@@ -6,49 +6,110 @@ import (
...
 
-			h.ServeHTTP(w, req)
+			h.ServeHTTP(w, req.WithContext(ctx))
 		})
 	}
 }
+
+func unmarshalResponse(resp *http.Response) (*identityResponse, error) {
+
+	var identityResp *identityResponse
```
Comment (mattrout92) [inline identity/identity.go]: consider naming the variables in the response part of function signature, e.g:

```go
func unmarshalResponse(resp *http.Response) (identityResp *identityResponse, err error) {
```

File: identity/identity.go [hunk 2]
```diff
@@ -6,49 +6,110 @@ import (
...
 
+					resp, err := cli.Do(ctx, zebReq)
+					if err != nil {
+						log.ErrorR(req, err, nil)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+
+					// Check to see if the florence user is authorised
+					if resp.StatusCode != http.StatusOK {
```
Comment (mattrout92) [inline identity/identity.go]: May want to log that an unexpected status code was returned

File: identity/identity.go [hunk 3]
```diff
@@ -6,49 +6,110 @@ import (
...
+	"context"
 	"github.com/ONSdigital/go-ns/log"
+	"encoding/json"
+	"io/ioutil"
 )
 
+//go:generate moq -out identitytest/http_client.go -pkg identitytest . HttpClient
+
 var zebedeeURL = os.Getenv("ZEBEDEE_URL")
+var florenceHeaderKey = "X-Florence-Token"
```
Comment (mattrout92) [inline identity/identity.go]: apart from zebedeeURL I think these are constants

File: identity/identity.go [hunk 4]
```diff
@@ -6,49 +6,110 @@ import (
...
+					zebReq, err := http.NewRequest("GET", zebedeeURL+"/identity", nil)
+					if err != nil {
+						log.ErrorR(req, err, nil)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+
+					if isUserReq {
+						zebReq.Header.Set(florenceHeaderKey, florenceToken)
+					} else if isServiceReq {
```
Comment (mattrout92) [inline identity/identity.go]: don't think you need the second if here, else will do

File: identity/identity.go (hunk 1 above)
Comment (CarlHembrough) [reply identity/identity.go]: Done 👍

File: identity/identity.go (hunk 2 above)
Comment (CarlHembrough) [reply identity/identity.go]: Done

File: identity/identity.go (hunk 3 above)
Comment (CarlHembrough) [reply identity/identity.go]: Changed to const

File: identity/identity.go (hunk 4 above)
Comment (CarlHembrough) [reply identity/identity.go]: Good point - done

**Pull request merged**